fedimint-api/src/db/notifications.rs

```diff
@@ -0,0 +1,230 @@
+use std::{
+    collections::hash_map::DefaultHasher,
+    hash::{Hash, Hasher},
+};
+
+use async_trait::async_trait;
+use bitvec::vec::BitVec;
+use tokio::sync::{futures::Notified, Notify};
+
+use super::Result;
+use super::{IDatabaseTransaction, PrefixIter};
+
+/// Number of buckets used for `Notifications`.
+const NOTIFY_BUCKETS: usize = 32;
+
+/// Manages notifications for the database
+///
+/// There are `NOTIFY_BUCKETS` number of `Notifies`.
+/// Each key is assigned a bucket based on its hash value.
+/// This will cause some false positives.
+#[derive(Debug)]
+pub struct Notifications {
+    buckets: Vec<Notify>,
+}
+
+impl Default for Notifications {
+    fn default() -> Self {
+        Self {
+            buckets: (0..NOTIFY_BUCKETS).map(|_| Notify::new()).collect(),
+        }
+    }
+}
+
+impl Notifications {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// This registers for notification when called.
+    ///
+    /// Then waits for the notification when .awaited.
+    ///
+    /// NOTE: This may some false positives.
+    pub fn register<K>(&self, key: &K) -> Notified
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.buckets.len() as u64)) as usize;
+        self.buckets[slot_idx].notified()
+    }
+
+    /// Notify a key.
+    ///
+    /// All the waiters for this keys will be notified.
+    pub async fn notify<K>(&self, key: &K)
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.buckets.len() as u64)) as usize;
+        self.buckets[slot_idx].notify_waiters();
+    }
+
+    /// Sends the notifications in the queue.
+    pub fn submit_queue(&self, queue: NotifyQueue) {
+        for bucket in queue.buckets.iter_ones() {
+            self.buckets[bucket].notify_waiters();
+        }
+    }
+}
+
+/// Save notifications to be sent on `submit_queue`.
+#[derive(Debug)]
+pub struct NotifyQueue {
+    buckets: BitVec,
+}
+
+impl Default for NotifyQueue {
+    fn default() -> Self {
+        Self {
+            buckets: BitVec::repeat(false, NOTIFY_BUCKETS),
+        }
+    }
+}
+
+impl NotifyQueue {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn add<K>(&mut self, key: &K)
+    where
+        K: Hash,
+    {
+        let mut hasher = DefaultHasher::new();
+        key.hash(&mut hasher);
+        let hash_value = hasher.finish();
+        let slot_idx = (hash_value % (self.buckets.len() as u64)) as usize;
+        self.buckets.set(slot_idx, true);
+    }
+}
+
+pub trait NotifyingTransaction<'a>: IDatabaseTransaction<'a> {
+    fn add_notification_key(&mut self, key: &[u8]);
+}
+
+pub struct NotifyingTransactionWrapper<'a> {
+    dbtx: Box<dyn IDatabaseTransaction<'a>>,
```

Comment:
It's not necessary, but you might consider trying `&dyn IDatabaseTransaction<'a>` like `IsolatedDatabaseTransaction` is doing, just to avoid another allocation.

Reply:
That is not possible because `commit` needs `Box<Self>` :/